src/main/java/org/truffleruby/core/format/TruncateStringNode.java

```diff
@@ -0,0 +1,28 @@
+package org.truffleruby.core.format;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import org.truffleruby.core.rope.RopeNodes;
+import org.truffleruby.core.string.RubyString;
+import org.truffleruby.core.string.StringNodes;
+//import org.truffleruby.core.string.StringNodesFactory;
+import org.truffleruby.core.string.StringNodesFactory.StringSubstringPrimitiveNodeFactory;
+
+public class TruncateStringNode extends FormatNode {
+
+    private final Object value;
+    private final Integer size;
+
+    public TruncateStringNode(Object value, Integer size) {
+        this.value = value;
+        this.size = size;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        RopeNodes.SubstringNode substringNode = RopeNodes.SubstringNode.create();
+        RubyString string = (RubyString) value;
+        string.setRope(substringNode.executeSubstring(string.rope, 0, this.size));
```

Comment:
@chrisseaton @eregon so I've learned that doing a substring of `size` bytes won't work for multibyte strings (pretty much what Chris pointed out in his comment).

I started looking for ways to substring `size` _characters_, and it seems like for that I need a bunch of Node objects:

* `StringSubstringPrimitiveNode` is where most of that logic lives
* If I wanted to copy some of that logic I'd also need `ByteIndexFromCharIndexNode`
* In general most of `StringSubstringPrimitiveNode` ends up being highly optimized for branches and then it's codegen'ed

I found it really hard to decouple that code from Node stuff. At the end it depends on the language configuration that is supposed to be in the top node.

How you'd recommend to approach this?

Reply:
Can you simply have a `@Child private StringSubstringPrimitiveNode substringNode;` in this class and use `StringSubstringPrimitiveNode#execute()`?
Like `org.truffleruby.core.string.StringNodes.GetIndexNode#slice(java.lang.Object, int, int)` does it in
https://github.com/oracle/truffleruby/blob/972a0521668532c070bdf9915e2524d3fefadb44/src/main/java/org/truffleruby/core/string/StringNodes.java#L576
https://github.com/oracle/truffleruby/blob/972a0521668532c070bdf9915e2524d3fefadb44/src/main/java/org/truffleruby/core/string/StringNodes.java#L785-L792
Or is the logic subtly different than `StringSubstringPrimitiveNode#execute`?

You could potentially also call to Ruby code here, but since `StringSubstringPrimitiveNode` is what's used for `String#[]` / `String#slice` that doesn't seem helpful here.